### PR TITLE
adding GreatExpectationsBigQueryOperator to bigquery.py

### PR DESCRIPTION
**Pull request by brian-lavery**
This is a request to add the GreatExpectationsBigQueryOperator to Airflow.  I'm probably bending protocol a bit because this is what I call a 'preliminary' pull request.  Before going to the effort of writing the tests for the new operator and submitting a more refined pull request, I need feedback about the viability of this operator being accepted into Airflow and any other major flaws I need to address.

I'll start with some background.  What is Great Expectations(GE)?  It's an open source python project that automates data QA testing and data profiling.  A user's data '[expectations](https://docs.greatexpectations.io/en/latest/reference/glossary_of_expectations.html)' can be laid out in a json file and GE will check those expectations against a table or query result by firing off a series of SQL queries.  Results of the checks are output in json and html files.  I've created a [short video](https://photos.app.goo.gl/PPHgCySA6iprVSWz5) to explain how GE and the operator work.

I'm looking for any feedback I can get but I have three immediate questions:
1. Does this sound like a contribution that would be accepted into Airflow?
2. Is there any special testing that needs to be done because the operator imports parts of the great expectations package, which has never been used before in Airflow?
3. Is it a problem that the operator connects to BigQuery in a somewhat unusual way compared with other operators?  The operator takes a BigQuery connection as a parameter but the actual connection to BigQuery is made inside of the GE code.  Therefore, the operator pulls some connection info from the hook but just passes that information to GE so it can make the connection.

Oh, and I reformatted the file in Pycharm so that's probably why there are more file differences compared with the previous version.  The real changes I made are:
1.  to the imports
2. Added the [new operator code](https://github.com/apache/airflow/blob/9a77b56d85aae295d220d619f363df2d585f42f2/airflow/providers/google/cloud/operators/bigquery.py#L2132) to the bottom of the file.

Thanks,
Brian

### CODE REVIEW

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 1]
```diff
@@ -20,26 +20,38 @@
...
 import hashlib
 import json
+import logging
+import random
 import re
+import string
+import time
 import uuid
-import warnings
+importwarnings
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py:33]:
```suggestion
import warnings
```

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 2]
```diff
@@ -158,15 +178,15 @@ class BigQueryCheckOperator(CheckOperator):
...
-        **kwargs,
+            self,
+            *,
+            sql: str,
+            gcp_conn_id: str = 'google_cloud_default',
+            bigquery_conn_id: Optional[str] = None,
+            use_legacy_sql: bool = True,
+            location: Optional[str] = None,
+            impersonation_chain: Optional[Union[str, Sequence[str]]] = None,
+            **kwargs,
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py:189]: This seems to be unrelated change

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 3]
```diff
@@ -231,17 +251,17 @@ class BigQueryValueCheckOperator(ValueCheckOperator):
...
+            *,
+            sql: str,
+            pass_value: Any,
+            tolerance: Any = None,
+            gcp_conn_id: str = 'google_cloud_default',
+            bigquery_conn_id: Optional[str] = None,
+            use_legacy_sql: bool = True,
+            location: Optional[str] = None,
+            impersonation_chain: Optional[Union[str, Sequence[str]]] = None,
+            **kwargs,
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py:264]: This seems to be unrelated change

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 4]
```diff
@@ -433,18 +453,18 @@ class BigQueryGetDataOperator(BaseOperator):
...
+            dataset_id: str,
+            table_id: str,
+            max_results: int = 100,
+            selected_fields: Optional[str] = None,
+            gcp_conn_id: str = 'google_cloud_default',
+            bigquery_conn_id: Optional[str] = None,
+            delegate_to: Optional[str] = None,
+            location: Optional[str] = None,
+            impersonation_chain: Optional[Union[str, Sequence[str]]] = None,
+            **kwargs,
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py:467]: This seems to be unrelated change

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 5]
```diff
@@ -616,32 +636,32 @@ def operator_extra_links(self):
...
+            query_params: Optional[list] = None,
+            labels: Optional[dict] = None,
+            priority: Optional[str] = 'INTERACTIVE',
+            time_partitioning: Optional[dict] = None,
+            api_resource_configs: Optional[dict] = None,
+            cluster_fields: Optional[List[str]] = None,
+            location: Optional[str] = None,
+            encryption_configuration: Optional[dict] = None,
+            impersonation_chain: Optional[Union[str, Sequence[str]]] = None,
+            **kwargs,
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py:664]: This seems to be unrelated change

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 6]
```diff
@@ -892,25 +912,25 @@ class BigQueryCreateEmptyTableOperator(BaseOperator):
...
+            bigquery_conn_id: str = 'google_cloud_default',
+            google_cloud_storage_conn_id: str = 'google_cloud_default',
+            delegate_to: Optional[str] = None,
+            labels: Optional[Dict] = None,
+            view: Optional[Dict] = None,
+            encryption_configuration: Optional[Dict] = None,
+            location: Optional[str] = None,
+            cluster_fields: Optional[List[str]] = None,
+            impersonation_chain: Optional[Union[str, Sequence[str]]] = None,
+            **kwargs,
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py:933]: This seems to be unrelated change as other added indentations in existing operators. You should be able to do this easily using pre-commit as described here:
https://github.com/apache/airflow/blob/master/STATIC_CODE_CHECKS.rst

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 7]
```diff
@@ -2109,3 +2127,268 @@ def execute(self, context: Any):
...
+                                Great Expectations Alert<br>
+                                One or more data expectations were not met in the {0} file. {1} 
+                           </div>
+                  </body>
+                </html>
+                '''.format(self.expectations_file_name, results)
+        send_email(self.email_to, 'expectations in ' + self.expectations_file_name + ' not met', email_content,
+                   files=None, cc=None, bcc=None,
+                   mime_subtype='mixed', mime_charset='us_ascii')
+        B
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py:2393]:
```suggestion

```
Not sure about it

**Review by turbaszek — CHANGES_REQUESTED** [review]
Hey @brian-lavery thanks for your contribution! I think this is an interesting add on to BigQuery family. However, I would like to ask you to remove all unrelated changes from this PR so we know what we should review 👌 

Answering your questions:
> 1. Does this sound like a contribution that would be accepted into Airflow?

I think yes

> 2. Is there any special testing that needs to be done because the operator imports parts of the great expectations package, which has never been used before in Airflow?

Yes we require unit tests for operators and an example DAG that can be easily run to test the operator in e2e way (I can help with that, but that's something we can add in the end). Having a docs will be awesome too as it will allow users to use the operators in easy way.

> 3. Is it a problem that the operator connects to BigQuery in a somewhat unusual way compared with other operators? The operator takes a BigQuery connection as a parameter but the actual connection to BigQuery is made inside of the GE code. Therefore, the operator pulls some connection info from the hook but just passes that information to GE so it can make the connection.

I think this sounds quite ok to me, but I will need to take a closer look at the details

**Comment by brian-lavery** [comment]
Hi @turbaszek .  Understood.  I'll cancel this pull request and submit another.

**Comment by ryw** [comment]
@mrrobby would you mind jumping in on this ticket?
